### PR TITLE
Add splitting guide

### PR DESCRIPTION
**Pull request by judithpatudith**
Add the splitting guide. 

- [PR to learn is here](https://github.com/hashicorp/learn/pull/727)
- [Preview build here](https://deploy-preview-727--hashicorp-learn.netlify.com/consul/developer-mesh/consul-splitting)

### CODE REVIEW

File: website/source/docs/guides/consul-splitting.md [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+The next configuration entry you need to add is the Service Resolver, which
+allows you to define how Consul’s service discovery selects service instances
+for a given service name.
+
+Service Resolvers allow you to filter for subsets of services based on
+information in the service registration. In this example, we are going to define
+the subsets “v1” and “v2” for the API service, based on its registered metadata.
+API service version 1 in the demo is already registered with the tags `v1` and
+service metadata `version:1`. When you register version 2 you will give it the
+tag `v2` and the metadata `version:2`. The `name` field is set to the name of
```
[the remark this answers is not in the document]
Comment (banks) [reply website/source/docs/guides/consul-splitting.md]: Hmm is this because the UI only shows tags right now and later you use that to show the difference?

I guess that's reasonable but maybe we should be super transparent about that so it's clear that tags are not actually needed for this splitting config to work?

@opihana @johncowen not sure if this is on you're radar although I feel like it's been mentioned in passing before but we should consider how the UI can use service meta in a similar way to how it surfaces tags currently to make this not a reason to still use tags going forward?

File: website/source/docs/guides/consul-splitting.md [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+Check that the service and its proxy have registered by checking for new `v2`
+tags next to the API service and API sidecar proxies in the Consul UI.
+
+### Configure Service Splitting - 50% Version 1, 50% Version 2
+
+Now that version 2 is running and registered, the next step is to gradually
+increase traffic to it by changing the weight of the v2 service subset in the
+service splitter configuration. Let’s increase the weight of the v2 service to
+50%. Remember; total service weight must equal 100, so you also reduce the
```
Comment (banks) [inline website/source/docs/guides/consul-splitting.md]: Can we change the language here and/or above to refer to the weight as a percentage? We explicitly designed this UX as percentages are easier to reason about than arbitrary sums/fractions. It's also important to note that these weights are floating point and can express up to 0.01% resolution. Something like:

> Remember; these are percentages so should sum to 100%, but note that you can use increments as small as 0.01%.

Comment (banks) [reply website/source/docs/guides/consul-splitting.md]: Is it also worth acknowledging that in production you'd likely start with a much smaller split until you gained confidence?

File: website/source/docs/guides/consul-splitting.md [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+  Service V2%
+$ curl localhost:9090
+Hello World
+###Upstream Data: localhost:9091###
+  Service V2%
+```
+
+Typically in a production environment, you would now remove the version 1
+service to release capacity in your cluster. Congratulations, you’ve now
+completed the deployment of version 2 of your service.
```
Comment (banks) [inline website/source/docs/guides/consul-splitting.md]: I saw "Clean Up" below and thought this was steps to remove the configs for splutting now we are done but it's not. Should we also include a section explaining how you would get configs back to a good state (i.e. remove the splitter again) once the v1 services are all decommissioned? We could either remove the splitter or remove the v1 branch ready for the next version release (would mean you could skip pre-creation of a splitter with 100% v2 etc.).

Also, how about calling the section below "Demo Clean Up" or something to make it clear this is not clean up you'd do in prod but just for this guide?

File: website/source/docs/guides/consul-splitting.md [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+![List of services in the Consul UI including Consul, and the web and API services with their proxies](/static/img/consul-splitting-services.png)
+
+Curl the Web endpoint to make sure that the whole application is running. The
+Web service will get a response from version 1 of the API service.
+
+```hcl
+$ curl localhost:9090
+Hello World
+###Upstream Data: localhost:9091###
+  Service V1%
```
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: Haha, I thought this too and then just glossed over it as I got more used to it. @nicholasjackson any insights?

File: website/source/docs/guides/consul-splitting.md [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+```
+
+The `kind` field denotes the type of configuration entry which you are defining;
+for this example, `service-defaults`. The `name` field defines which service the
+service-defaults configuration entry applies to. (The value of this field must
+match the name of a service registered in Consul, in this example, `api`.) The
+`protocol` is `http`.
+
+To apply the configuration, you can either use the Consul CLI or the API. In
+this example we’ll use the configuration entry endpoint of the HTTP API, which
```
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: Oh interesting, I didn't know that we had a preference. 

Usually, I default to the HTTP API because I assume that when users have big datacenters they might not want to log into the right node to register a new service or check, etc. 

Not sure what a real-life workflow looks like, but we can definitely default to the CLI from now on 😄

File: website/source/docs/guides/consul-splitting.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: FWIW the behavior that I originally expected was that you could filter directly on tags. I haven't tried this, so maybe it's possible and I just don't know?

File: website/source/docs/guides/consul-splitting.md [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+- Service Defaults for the API service to set the protocol to HTTP.
+- Service Splitter which defines the traffic split between the service subsets.
+- Service Resolver which defines which service instances are version 1 and 2.
+
+### Configuring Service Defaults
+
+Traffic splitting requires that the upstream application uses HTTP, because
+splitting happens on layer 7 (on a request by request basis). You will tell
+Consul that your upstream service uses HTTP by setting the protocol in a
+“service defaults” configuration entry for the API service. This configuration
```
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: I lean toward using plain language when you're talking about a Consul item and the code snippet format when referring to the field, or a specific file or directory. In this case, I think the noun would be "configuration entry" and the adjective describing it is "service defaults". Because it's modifying a noun it should be hyphenated. Ex: "service-defaults configuration entry". Hope that makes sense, and your instinct that this is weird is correct. Fixing it now :P

File: website/source/docs/guides/consul-splitting.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: I was wondering this too and just got used to it and forgot! @nicholasjackson any insight?

File: website/source/docs/guides/consul-splitting.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (banks) [reply website/source/docs/guides/consul-splitting.md]: Yeah you definitely can filter just on tags it would look like `v1 in Service.Tags` in the subset filter string (not tested but pretty sure that's right).

I think the question is whether we want to promote tag usage as we'd kinda like to deprecate it but currently the UI only shows tags... I'm OK only using tags now if it makes this guide simpler but would be nice to update to use meta only once we have a UI solution that allows meta-only service instances to have some meta visible easily in the UI?

File: website/source/docs/guides/consul-splitting.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (banks) [reply website/source/docs/guides/consul-splitting.md]: Yeah I think HTTP does make some sense for that reason but I think our CLI usually offers a better experience so we should maybe try to steer towards using that where possible. In other words, we have more control over the UX with the CLI (c.f. creating intentions with `consul intentions create -allow foo bar` vs trying to do that with JSON and curl.)

It's def possible to use consul CLI from a bastion host or similar but we perhaps don't spell that out anywhere. Maybe we should? By default it connects to localhost because it expects every machine to be running an agent, but it never needs to be run from the actual server hosts and if it's on a host with no agent you can point it to another host via ENV of args easily enough.

What do you think?

[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: I think having a guide about connecting the CLI would be great. Putting it on our to-do list now so it doesn't get lost. 

The one other advantage that I just thought of with giving the HTTP API example is that I assume people are going to want to automate most stuff. As an example, Nic's original blog was really canary focused, and my understanding is that sometimes people will automate those based on pre-defined thresholds. I assume organizations that are actually that sophisticated are few and far between though.

In any case, I'll replace the HTTP examples here with CLI instead 😄 Nic updated the application a little bit so this will give me an excuse to retest with the updates.

File: website/source/docs/guides/consul-splitting.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (johncowen) [reply website/source/docs/guides/consul-splitting.md]: Hey folks 👋 

TLDR; (it got long 😆 ) In my mind tags are for humans, I think we should recommend using meta data for splitting. Metadata is surfaced in the UI but we need to make it easier to get to.

---

Yeah tags are a funny one, and they are becoming more 'funny peculiar' as we move into the discovery chain/resolving work.

So a bit of background: Specifically in the UI we 'rollup' tags on your Services on the service listing page, rollup referring to the fact that we aggregate all the tags from all the Service Instances of a particular Service into the same list. That means that if you have some Service Instances with the tag `v1` and some with the tag `v2`, when you look at the Service listing the Service has _both_ `v1` and `v2`.

This isn't a UI thing exactly as it comes from the API like this, i.e. we don't aggregate this on the frontend. For me this is sometimes confusing, yet kind of 'correct'.

We've spoken quite a lot about using/not using 'rollups' (aggregating Service Instance data into the Service) for surfacing things like this. Ideally I'd fall down on the side of moving away from rolling-up unless we can make it so the user can explicitly say 'this _entire_ Service has the tag v1', this maybe achievable via centralized config in a very distant future. Also, 'ideally' isn't something we can always do as it breaks workflows for others.

---

My 2 cents on what we could do: As @banks mentioned we are kind of 'deprecating' tags, tags being just a type of metadata that helps to describe an item and makes it easy to find again at a later date. I personally feel that we should use Meta itself for anything 'technical/machine' related moving forwards which I think is what @banks is suggesting, and _if anything_ use Tags for 'humans' to enable them to find things (getting very specific this is what a tag is for, making it easier for humans to find things) either via the UI or something else. So for this exact case ^ in the docs, maybe just use Meta in the examples but mention below as an aside that you can filter on other things such as Tags (and whatever else folks can filter on)? 

From a UI standpoint, we recently added Meta to the UI, but you can only see this if click into the Service Instance and then into a Meta Data tab (i.e. we don't 'roll them up' currently like we do with tags), this is still 'correct' but less confusing when working with the discovery chain - and is four clicks away from the Service listing! 🙀 

I would imagine people immediately reach for tags to further categorise Services partly because they are more visible in the UI, so I definitely think we should make Meta much easier to reach and browse in the UI and I think there are a few options that we've spoken about before and I think we should see if we can pick these discussions back up. @opihana , @jnwright I'm gonna add this to our agenda for our next catch up but obvs feel free to pitch in here also.

File: website/source/docs/guides/consul-splitting.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: Nic says that this has something to do with the command output because we're using curl. Sounds like it would be annoying to hide, but we can explain it briefly if you think it's worth it. My inclination is to leave it as is, since I don't think it's going to change people's perception of Consul functionality.

**Comment by banks** [comment]
Hmm, I don't think it's huge but I do feel like it confused me especially
given that we are talking about splitting percentages. I thought at first
it was trying to be smart and display the current split percentage but
realised that was strange.

Does it go away with `curl -q`?

File: website/source/docs/guides/consul-splitting.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: @Paul I think your last comment is displaying strangely. Copying it back here for posterity: 

> Hmm, I don't think it's huge but I do feel like it confused me especially
given that we are talking about splitting percentages. I thought at first
it was trying to be smart and display the current split percentage but
realised that was strange. Does it go away with `curl -q`?

🤔 It doesn't go away. I'll google around more and see what I can do

[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: It seems like the `%` is something zsh adds because there is no new line in the output that the API service returns. It looks even more confusing if you use bash; it appends the next command prompt on to the end of the output from the service: `  Service V1bash-3.2$`.

I think the best fix is to add a newline to the service output. I submitted an issue, and If I have time I can dig around more and try to fix it myself. https://github.com/hashicorp/consul-demo-traffic-splitting/issues/4

[the remark this answers is not in the document]
Comment (banks) [reply website/source/docs/guides/consul-splitting.md]: Oooh good detection skillz.

In that case I'd be tempted just to remove it from the guide and publish. We can add a new line to the code maybe but I think it's less confusing if people see their own shell behaviour than wonder what this is reading the guide.

If you want a solution that looks completely authentic you could add `-w "\m"` to the curl command which [will add the newline for you](https://stackoverflow.com/questions/12849584/automatically-add-newline-at-end-of-curl-response-body). But I don't personally think that's necessary for our users to understand the guide or even follow along.

File: website/source/docs/guides/consul-splitting.md [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+to centrally configure the services and Envoy proxies. There are three
+configuration entries you need to create to enable traffic splitting:
+
+- Service Defaults for the API service to set the protocol to HTTP.
+- Service Splitter which defines the traffic split between the service subsets.
+- Service Resolver which defines which service instances are version 1 and 2.
+
+### Configuring Service Defaults
+
+Traffic splitting requires that the upstream application uses HTTP, because
```
[the remark this answers is not in the document]
Comment (judithpatudith) [reply website/source/docs/guides/consul-splitting.md]: fixed in the learn repo

**Comment by judithpatudith** [comment]
Ok, I think I got all the comments. If you'd like to see the test build it's [here](https://deploy-preview-727--hashicorp-learn.netlify.com/consul/developer-mesh/consul-splitting).

**Comment by i0rek** [comment]
Is this still relevant?
